### ISSUE TITLE
proposal: path/filepath: Deprecate Walk

### ISSUE THREAD

**Issue by earthboundkid** [golang/go#69604]
### Proposal Details

filepath.WalkDir was added in Go 1.16. It's past time to go ahead and deprecate filepath.Walk, which is less efficient.

**Comment by robpike** [comment]
Define 'deprecate'.

**Comment by ianlancetaylor** [comment]
I'm not sure it's useful to deprecate a function that doesn't have an exact equivalent that people can simply switch to.  The docs for `filepath.Walk` already say "Walk is less efficient than [WalkDir](https://pkg.go.dev/path/filepath#WalkDir), introduced in Go 1.16, which avoids calling os.Lstat on every visited file or directory."

**Comment by earthboundkid** [comment]
> Define 'deprecate'.

Add `// Deprecated: Use [WalkDir].` to the doc string, so my editor puts a squiggly line under accidental uses of Walk. I am aware that WalkDir exists and is more efficient but accidentally used Walk instead last week.

**Comment by ianlancetaylor** [comment]
Yes, you can update, but the update can't be easily done mechanically, and I don't think the performance advantage here is sufficient reason to push people to modify existing working code.

I think what you are looking for is something more like "hint when writing new code that a different choice would be better".  Our current uses of deprecation apply to existing code as well.  As far as I know we don't have that kind of hint at present.

**Comment by earthboundkid** [comment]
I don't think the expectation is that there's always a trivial find-and-replace for deprecations. For example, [`zip.FileHeader.ModifiedTime`](https://pkg.go.dev/archive/zip#FileHeader) is a `uint16` that is deprecated in favor of a `time.Time` Modified field. Presumably fixing that requires ripping out any code that uses the `uint16` as a number. 

My expectation is that a large percentage of the changes from Walk to WalkDir would just need to change from fs.FileInfo to fs.DirEntry in the callback, but the ones that need more than that are fairly easy to catch and fix when the type check fails. I think it actually could be done mechanically with a type aware rewrite tool.

**Comment by ianlancetaylor** [comment]
I'm not sure it's quite analogous.  `zip.FileHeader.ModifiedTime` is deprecated because it doesn't have the same precision.  Switching to `Modified` changes from 2-second precision to 1-second precision. So `Modified` is really better.

`Walk` and `WalkDir` are nearly the same.  There isn't much of a reason to change existing code.

To be clear I certainly agree that new code should use `WalkDir`.  I just don't think that deprecation as implemented today is necessarily the best path to make that happen.

**Comment by apocelipes** [comment]
Rather than marking a function as deprecated, I'd like "go vet" or gopls to give me some hints like "WalkDir performs better, you should use this" or something like that. And if you add this hint, the same thing should be applied to `os.(*File).Readdir` too.

**Comment by robpike** [comment]
The comments already say that. Go vet's guildelines (see the README) have no scope to do suggest faster code. Whether gopls's job includes repeating all the information in the comments is up to its authors. I don't believe it does, though.

**Comment by apparentlymart** [comment]
Reading the discussion so far, I agree that it seems like the _most ideal_ answer here would be a new classification that is softer than "deprecated" but still machine-readable so that tools can react to it. However, I also have a feeling that discussing such a thing here is taking the issue off-topic. :thinking: 

At the risk of straying _further_ off-topic :grimacing: ...

The idea of this being a `gopls` thing actually resonates with me because its code completion hints are _very much_ relevant only when writing new code, and not when reading existing code. Is this issue describing a use-case for a way to declare that a particular symbol should be excluded from (or deprioritized in) code completion?

Of course that's not an answer for everyone, but I will admit that when I'm writing new code I do tend to "follow my nose" by selecting the most likely candidate from the code completion list, rather than carefully reviewing the documentation for each function. Having `Walk` either not appear at all, or always appear after `WalkDir` in the choice list, would make me more likely to select it.

**Comment by seankhliao** [comment]
I think as above, Deprecation is too strong for filepath.Walk.
